src/DotLiquid/Util/StrFTime.cs

```diff
@@ -16,87 +21,201 @@ public static class StrFTime
+                    x.Groups["width"].Captures.Cast<Capture>().Select(y => (int?)Convert.ToInt32(y.Value)).FirstOrDefault(),
+                    x.Groups["directive"].Captures.Cast<Capture>().Select(y => y.Value).FirstOrDefault(),
+                    dateTime
+                    ));
+        }
 
-            int n = 0;
+        /// <summary>
+        /// Formats a date using a ruby date format string
+        /// </summary>
+        /// <param name="dateTimeOffset">date-time object to be formatted</param>
+        /// <param name="format">the required format</param>
+        /// <returns>a string version of date-time matching pattern.</returns>
+        public static string ToStrFTime(this DateTimeOffset dateTimeOffset, string format)
+        {
+            return Regex.Replace(input: format, pattern: PATTERN_REGEX,
+                evaluator: x => StrFTimeMatchEvaluator(
+                    x.Groups[0].Value,
+                    x.Groups["flag"].Captures.Cast<Capture>().Select(y => y.Value).ToList(),
+                    x.Groups["width"].Captures.Cast<Capture>().Select(y => (int?)Convert.ToInt32(y.Value)).FirstOrDefault(),
+                    x.Groups["directive"].Captures.Cast<Capture>().Select(y => y.Value).FirstOrDefault(),
+                    dateTimeOffset
+                    ));
+        }
 
-            while (n < pattern.Length)
-            {
-                string s = pattern.Substring(n, 1);
 
-                if (n + 1 >= pattern.Length)
-                    output += s;
-                else
-                    output += s == "%"
-                        ? Formats.ContainsKey(pattern.Substring(++n, 1)) ? Formats[pattern.Substring(n, 1)].Invoke(dateTime) : "%" + pattern.Substring(n, 1)
-                        : s;
-                n++;
-            }
+        /// <summary>
+        /// Formats a date-time for a single specifier within the requested format. For ease of processing
+        /// the specifier is broken down into it's consituent parts of: flags, width, directive.
+        ///
+        /// A specifier consists of a percent (%) character, zero or more flags, optional minimum field width and a conversion directive as follows:
+        /// <param name="orig">the entire specifier, such as %Y for a 4-digit year</param>
+        /// <param name="flags">zero or more flags to change formatting, such as '0' (zero padding).</param>
+        /// <param name="width">optional minimum field width</param>
+        /// <param name="directive">The required data value, such as Y for a 4-digit year</param>
+        /// <param name="obj">the source date-time object</param>
+        /// </summary>
+        private static String StrFTimeMatchEvaluator(String orig, IEnumerable<String> flags, int? width, String directive, object obj)
+        {
+            var result = orig;
+            directive = PreProcessDirective(directive, flags, width);
+
+            bool isDateTimeOffset = obj is DateTimeOffset;

```

Comment:
use pattern matching

var isDateTimeOffset = obj is DateTimeOffset dateTimeOffset;

then you don't need to cast again when the pattern matches, instead use dateTimeOffset below

Reply:
This triggered a CS0165 compiler error for netstandard2.0, to resolve I created a separate named pattern in the if and else blocks.